packages/zod-validator/src/index.ts

```diff
@@ -46,10 +46,26 @@ export const zValidator = <
 ): MiddlewareHandler<E, P, V> =>
   // @ts-expect-error not typed well
   validator(target, async (value, c) => {
-    const result = await schema.safeParseAsync(value)
+    let validatorValue = value
+
+    // in case where our `target` === `header`, Hono parses all of the headers into lowercase.
+    // this might not match the Zod schema, so we want to make sure that we account for that when parsing the schema.
+    if (target === 'header' && schema instanceof ZodObject) {
+      // create an object that maps lowercase schema keys to lowercase
+      const schemaKeys = Object.keys(schema.shape)
+      const caseInsensitiveKeymap = Object.fromEntries(
+        schemaKeys.map((key) => [key.toLowerCase(), key])
```

Comment:
It is a very very slight difference, but `toLowerCase()` takes longer to execute. Therefore, this PR will cause performance degradation. We would like to avoid that as much as possible. Also, we may need to add the same procedure to other validators if we will add this feature for the Zod Validator.

What do you think about these?

We can merge this PR if we can allow for these pain points.

But if not so, for example, could we encourage documents to use only lowercase? We wrote about it in the `Validation` section of the website. https://hono.dev/docs/guides/validation

Reply:
I think the issue with documentation is it's very easy to miss. For example, I've been using Hono for over 2 years at this point and never knew headers are casted to lower case if you retrieve them through `c.req.headers`...

As a runtime-performant solution I might have an idea. What if restrict the schema keys for `headers` target to only lowercase characters? Is that something that could work?

Reply:
> What if restrict the schema keys for `headers` target to only lowercase characters using TypeScript? Is that something that could work? This way it would just let you know in an IDE that you have a problem with your headers validation

This is a super interesting idea. We have to find good ways.